rfc/rfc-82/rfc-82.md

```diff
@@ -0,0 +1,327 @@
+| 4 | S1 | S1 | S1 | No | S1 | No schema evolution |
+| 5 | S1 | S1 | S2 | No | S2 | Schema evolution in current transaction |
+| 6 | S1 | S2 | S1 | No | S2 | Backwards compatibility handles it |
+| 7 | S1 | S2 | S2 | No | S2 | Concurrent evolution to same schema |
+| 8 | S1 | S2 | S3 | Yes | N/A (throws exception) | Concurrent evolution to different schemas. Same as case 2, for future optimization we should consider evolvibility from S2 to S3 |
+
+For timeline graph of each case please refer appendix.
+
+Notes:
+- S1, S2, S3 represent different schemas. The proposed solution does not introduce any new assumptions on compatibility among those schemas. But if "Table Schema when Txn Start" is schema X and Table Schema when Txn Validates is Y, it naturally means Y is compatible and is evolved from X as it is guaranteed by today's implementation.
+- 3 schemas to consider:
+  + The **table schema** from the last committed txn when the current txn starts.
+  + The **table schema** from the last committed txn when the current txn validates.
+  + The **writer schema** that the current txn uses.
+- "Not exists" means the table is empty and **table schema** is not set.
+
+## Implementation
+
+The proposed implementation involves the following key changes:
+
+1. Enhance the `TransactionUtils` class to include schema conflict detection:
+   - Add a new interface `SchemaConflictResolutionStrategy` to check for schema conflicts.
+   - Add an implementation to this new interface `SimpleSchemaConflictResolutionStrategy` which implements the table above.
+
+2. Schema conflict detection logic:
+   - Follow the graph as explained above.
+
+sudo code:
+```
+// Initialize schemas
+schemaOfTxn = current transaction schema
+schemaAtTxnStart = schema at transaction start (if available)
+schemaAtTxnValidation = schema at transaction validation (if available)
+
+// Case 1: First commit ever
+if schemaAtTxnValidation is null:
+    return schemaOfTxn
+
+// Case 2, 3: Second commit, first one committed during read-write phase
+if schemaAtTxnStart is null:
+    if schemaOfTxn != schemaAtTxnValidation:
+        throw ConcurrentSchemaEvolutionError
+    return schemaOfTxn
+
+// Case 8: Multiple commits, potential concurrent schema evolution
+if schemaAtTxnStart != schemaOfTxn AND
+   schemaAtTxnStart != schemaAtTxnValidation AND
+   schemaOfTxn != schemaAtTxnValidation:
+    throw ConcurrentSchemaEvolutionError
+
+// Compatible case 4,5
+if transaction start instant == transaction validation instant:
+    return schemaOfTxn
+
+// Compatible case 7
+if schemaAtTxnStart == schemaAtTxnValidation:
+    return schemaOfTxn
+
+// Compatible case 6
+return schemaAtTxnValidation
+```
```

Comment:
```suggestion
```
// Initialize schemas
writerSchemaOfTxn = writer schema of the current transaction
tableSchemaAtTxnStart = table schema at transaction start (if available)
tableSchemaAtTxnValidation = table schema at transaction validation (if available)

// Case 1: First commit ever
if tableSchemaAtTxnValidation is null:
    return writerSchemaOfTxn

// Case 2, 3: Second commit, first one committed during read-write phase
if tableSchemaAtTxnStart is null:
    if writerSchemaOfTxn != tableSchemaAtTxnValidation:
        throw ConcurrentSchemaEvolutionError
    return writerSchemaOfTxn

// Case 8: Multiple commits, potential concurrent schema evolution
if tableSchemaAtTxnStart != writerSchemaOfTxn AND
   tableSchemaAtTxnStart != tableSchemaAtTxnValidation AND
   writerSchemaOfTxn != tableSchemaAtTxnValidation:
    throw ConcurrentSchemaEvolutionError

// Compatible case 4,5
if transaction start instant == transaction validation instant:
    return writerSchemaOfTxn

// Compatible case 7
if tableSchemaAtTxnStart == tableSchemaAtTxnValidation:
    return writerSchemaOfTxn

// Compatible case 6
return tableSchemaAtTxnValidation
```
```

Reply:
done

Reply:
done